common/datavalues2/src/columns/column.rs

```diff
@@ -0,0 +1,157 @@
+
+use common_arrow::arrow::array::ArrayRef;
+use common_arrow::arrow::bitmap::Bitmap;
+use common_arrow::arrow::bitmap::MutableBitmap;
+use common_exception::Result;
+
+use crate::prelude::*;
+use crate::BooleanColumn;
+use crate::DataTypePtr;
+use crate::DataValue;
+use crate::NullColumn;
+use crate::TypeID;
+
+pub type ColumnRef = Arc<dyn Column>;
+pub trait Column: Send + Sync {
+    fn as_any(&self) -> &dyn Any;
+    /// Type of data that column contains. It's an underlying physical type:
+    /// UInt16 for Date, UInt32 for DateTime, so on.
+    fn data_type_id(&self) -> TypeID {
+        self.data_type().data_type_id()
+    }
+    fn data_type(&self) -> DataTypePtr;
+
+    fn is_nullable(&self) -> bool {
+        false
+    }
+
+    fn is_const(&self) -> bool {
+        false
+    }
+
+    fn len(&self) -> usize;
+    /// whether the array is empty
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    fn null_at(&self, _row: usize) -> bool {
+        false
+    }
+
+    /// If the only value column can contain is NULL.
+    fn only_null(&self) -> bool {
+        false
+    }
+
+    /// Returns (is_all_null,  Option bitmap)
+    fn validity(&self) -> (bool, Option<&Bitmap>) {
+        (false, None)
+    }
+
+    fn memory_size(&self) -> usize;
+    fn as_arrow_array(&self) -> ArrayRef;
+    fn slice(&self, offset: usize, length: usize) -> ColumnRef;
+
+    // Copies each element according offsets parameter.
+    // (i-th element should be copied offsets[i] - offsets[i - 1] times.)
+    fn replicate(&self, offsets: &[usize]) -> ColumnRef;
+
+    fn convert_full_column(&self) -> ColumnRef;
+
+    /// # Safety
+    /// Assumes that the `index` is smaller than size.
+    unsafe fn get_unchecked(&self, index: usize) -> DataValue;
+
+    /// # Safety
+    /// Assumes that the `index` is smaller than size.
+    unsafe fn get_u64_unchecked(&self, index: usize) -> Result<u64> {
+        let value = self.get_unchecked(index);
+        DFTryFrom::try_from(&value)
+    }
+
+    /// # Safety
+    /// Assumes that the `index` is smaller than size.
+    unsafe fn get_i64_unchecked(&self, index: usize) -> Result<i64> {
+        let value = self.get_unchecked(index);
+        DFTryFrom::try_from(&value)
+    }
+
+    /// # Safety
+    /// Assumes that the `index` is smaller than size.
+    unsafe fn get_string_unchecked(&self, index: usize) -> Result<Vec<u8>> {
+        let value = self.get_unchecked(index);
+        DFTryFrom::try_from(value)
+    }
+}
+
+pub trait IntoColumn {
+    fn into_column(self) -> ColumnRef;
+    fn into_nullable_column(self) -> ColumnRef;
+}
+
+// No nullable
+// We should wrap the nullable by ourselves
+impl IntoColumn for ArrayRef {
+    fn into_column(self) -> ColumnRef {
+        use TypeID::*;
+        let data_type: DataTypePtr = from_arrow_type(self.data_type());
+        match data_type.data_type_id() {
+            Nullable | Null => Arc::new(NullColumn::from_arrow_array(self.as_ref())),
```

Comment:
Maybe `Nullable => unimplemented!()` is better?

Reply:
Right, it's a small fix. Will be addressed in another pr.